crates/ide/src/references/rename.rs

```diff
@@ -171,72 +168,97 @@ fn find_definition(
 }
 
 fn source_edit_from_references(
-    sema: &Semantics<RootDatabase>,
+    _sema: &Semantics<RootDatabase>,
     file_id: FileId,
     references: &[FileReference],
+    def: Definition,
     new_name: &str,
 ) -> (FileId, TextEdit) {
     let mut edit = TextEdit::builder();
     for reference in references {
-        let mut replacement_text = String::new();
-        let range = match reference.kind {
-            ReferenceKind::FieldShorthandForField => {
-                mark::hit!(test_rename_struct_field_for_shorthand);
-                replacement_text.push_str(new_name);
-                replacement_text.push_str(": ");
-                TextRange::new(reference.range.start(), reference.range.start())
-            }
-            ReferenceKind::FieldShorthandForLocal => {
-                mark::hit!(test_rename_local_for_field_shorthand);
-                replacement_text.push_str(": ");
-                replacement_text.push_str(new_name);
-                TextRange::new(reference.range.end(), reference.range.end())
-            }
-            ReferenceKind::RecordFieldExprOrPat => {
-                mark::hit!(test_rename_field_expr_pat);
-                replacement_text.push_str(new_name);
-                edit_text_range_for_record_field_expr_or_pat(
-                    sema,
-                    FileRange { file_id, range: reference.range },
-                    new_name,
-                )
-            }
-            _ => {
-                replacement_text.push_str(new_name);
-                reference.range
-            }
+        let (range, replacement) = match &reference.name {
+            ast::NameLike::Name(_) => (None, format!("{}", new_name)),
+            ast::NameLike::NameRef(name_ref) => source_edit_from_name_ref(name_ref, new_name, def),
+            ast::NameLike::Lifetime(_) => (None, format!("{}", new_name)),
         };
-        edit.replace(range, replacement_text);
+        // FIXME: Some(range) will be incorrect when we are inside macros
```

Comment:
Basically when more than just the name gets renamed we will get a `Some(range)` for what to replace, this won't work well with ranges inside macros, I'm not sure how to handle that. Best would be to just not allow special renaming, like shorthand related stuff, to be done in macro expanded files I guess?

Reply:
see https://github.com/rust-analyzer/rust-analyzer/pull/7358#discussion_r562707303